ext/reflection/php_reflection.c

```diff
@@ -2106,7 +2131,136 @@ ZEND_METHOD(ReflectionFunctionAbstract, getNumberOfRequiredParameters)
+/* {{{ Returns whether a parameter exists or not */
+ZEND_METHOD(ReflectionFunctionAbstract, hasParameter)
+{
+    reflection_object *intern;
+    zend_function *fptr;
+    zend_string *arg_name = NULL;
+    zend_long position;
+    uint32_t num_args;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_STR_OR_LONG(arg_name, position)
+    ZEND_PARSE_PARAMETERS_END();
+
+    GET_REFLECTION_OBJECT_PTR(fptr);
+
+    num_args = fptr->common.num_args;
+
+    if (fptr->common.fn_flags & ZEND_ACC_VARIADIC) {
+        num_args++;
+    }
+
+    if (!num_args) {
+        RETURN_FALSE;
+    }
+
+    if (arg_name != NULL) {
+        if (ZSTR_LEN(arg_name) == 0) {
+            zend_argument_value_error(1, "must not be empty");
+            RETURN_THROWS();
+        }
+
+        if (get_parameter_position(fptr, arg_name, num_args) > -1) {
+            RETURN_TRUE;
+        }
+
+        RETURN_FALSE;
+    } else {
+        if (position < 0) {
+            zend_argument_value_error(1, "must be greater than or equal to 0");
+            RETURN_THROWS();
+        }
+
+        RETURN_BOOL(position < num_args);
+    }
+}
+/* }}} */
+
+/* {{{ Returns a parameter specified by its name */
+ZEND_METHOD(ReflectionFunctionAbstract, getParameter)
+{
+    reflection_object *intern;
+    zend_function *fptr;
+    zend_string *arg_name = NULL;
+    zend_long position;
+    struct _zend_arg_info *arg_info;
+    uint32_t num_args;
+    zval reflection;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_STR_OR_LONG(arg_name, position)
+    ZEND_PARSE_PARAMETERS_END();
+
+    GET_REFLECTION_OBJECT_PTR(fptr);
+
+    num_args = fptr->common.num_args;
+    arg_info = fptr->common.arg_info;
+
+    if (fptr->common.fn_flags & ZEND_ACC_VARIADIC) {
+        num_args++;
+    }
+
+    if (num_args < 1) {
+        if (fptr->common.scope) {
+            _DO_THROW_EX("Method %s::%s() has no parameters", ZSTR_VAL(fptr->common.scope->name), ZSTR_VAL(fptr->common.function_name));
+        } else {
+            _DO_THROW_EX("Function %s() has no parameters", ZSTR_VAL(fptr->common.function_name));
+        }
+
+        RETURN_THROWS();
+    }
+
+    if (arg_name != NULL) {
+        if (ZSTR_LEN(arg_name) == 0) {
+            zend_argument_value_error(1, "must not be empty");
+            RETURN_THROWS();
+        }
+
+        position = get_parameter_position(fptr, arg_name, num_args);
+
+        if (position == -1) {
+            if (fptr->common.scope) {
+                _DO_THROW_EX("Method %s::%s() has no parameter named \"%s\"",
+                             ZSTR_VAL(fptr->common.scope->name), ZSTR_VAL(fptr->common.function_name), ZSTR_VAL(arg_name));
+            } else {
+                _DO_THROW_EX("Function %s() has no parameter named \"%s\"",
+                             ZSTR_VAL(fptr->common.function_name), ZSTR_VAL(arg_name));
+            }
+            RETURN_THROWS();
+        }
+    } else {
+        if (position < 0) {
```

Comment:
Wait, do these functions use zero index for parameters? I think it's not a good idea, since error messages use a 1-index (`Argument #1 ...`)

Reply:
I know that `getParameters()` returns a zero-index array, so it's not exactly clear which indexing is the most expected for people. For me, it's the 1-based one. 🤔

Reply:
My opinion is that we should use zero index to be consistent with getPosition and the parameter array.
I read "Argument #1 ..." in my head as "argument *number* 1", not "argument *index* 1".
Anyway, this needs discussion on the ML as said before.

Reply:
> My opinion is that we should use zero index to be consistent with getPosition and the parameter array. I read "Argument #1 ..." in my head as "argument _number_ 1", not "argument _index_ 1". Anyway, this needs discussion on the ML as said before.

I agree, as we're used to dealing with 0-index arrays.

Shall I just continue as is for now, and then send something to the ML? What specifically should I be mentioning? This PR and the whole discussion over index vs position?

Reply:
I would prefer a discussion on the ML before anything is done here.

And the discussion should be, here are two new helper methods which take either the name of a parameter or its position. However there is contention between should this be 1 indexed or 0 indexed due to the existing API.

Reply:
> I would prefer a discussion on the ML before anything is done here.
> 
> And the discussion should be, here are two new helper methods which take either the name of a parameter or its position. However there is contention between should this be 1 indexed or 0 indexed due to the existing API.

Have started one on there @Girgias, sorry about the delay.

https://externals.io/message/123251